config/store.go

```diff
@@ -52,15 +41,249 @@ type Store interface {
+	}
+
+	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
+		return nil, errors.Wrap(err, "failed to load on save")
+	}
+
+	return oldCfg, nil
+}
+
+func (s *Store) loadLockedWithOld(oldCfg *model.Config, unlockOnce *sync.Once) error {
+	configBytes, err := s.backingStore.Load()
+	if err != nil {
+		return err
+	}
+
+	var loadedConfig model.Config
+	if len(configBytes) != 0 {
+		if err = json.Unmarshal(configBytes, &loadedConfig); err != nil {
+			return jsonutils.HumanizeJsonError(err, configBytes)
+		}
+	}
+
+	loadedConfig.SetDefaults()
+
+	s.configNoEnv = loadedConfig.Clone()
+	fixConfig(s.configNoEnv)
+
+	loadedConfig = *applyEnviromentMap(&loadedConfig, GetEnviroment())
+
+	fixConfig(&loadedConfig)
+
+	if err := loadedConfig.IsValid(); err != nil {
+		return errors.Wrap(err, "invalid config")
+	}
+
+	// Apply changes that may have happened on load to the backing store.
+	oldCfgBytes, err := json.Marshal(oldCfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal old config")
+	}
+	newCfgBytes, err := json.Marshal(loadedConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal loaded config")
+	}
+	if len(configBytes) == 0 || !bytes.Equal(oldCfgBytes, newCfgBytes) {
+		if err := s.backingStore.Set(s.configNoEnv); err != nil {
+			if !errors.Is(err, ErrReadOnlyConfiguration) {
+				return errors.Wrap(err, "failed to persist")
+			}
+		}
+	}
+
+	s.config = &loadedConfig
+
+	unlockOnce.Do(s.configLock.Unlock)
+
+	s.invokeConfigListeners(oldCfg, &loadedConfig)
+
+	return nil
+}
+
+// Load updates the current configuration from the backing store, possibly initializing.
+func (s *Store) Load() error {
+	s.configLock.Lock()
+	var unlockOnce sync.Once
+	defer unlockOnce.Do(s.configLock.Unlock)
+
+	oldCfg := s.config.Clone()
+
+	return s.loadLockedWithOld(oldCfg, &unlockOnce)
+}
+
+// GetFile fetches the contents of a previously persisted configuration file.
+// If no such file exists, an empty byte array will be returned without error.
+func (s *Store) GetFile(name string) ([]byte, error) {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.backingStore.GetFile(name)
+}
+
+// SetFile sets or replaces the contents of a configuration file.
+func (s *Store) SetFile(name string, data []byte) error {
+	s.configLock.Lock()
+	defer s.configLock.Unlock()
+	return s.backingStore.SetFile(name, data)
+}
+
+// HasFile returns true if the given file was previously persisted.
+func (s *Store) HasFile(name string) (bool, error) {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
+	return s.backingStore.HasFile(name)
+}
+
+// RemoveFile removes a previously persisted configuration file.
+func (s *Store) RemoveFile(name string) error {
+	s.configLock.Lock()
+	defer s.configLock.Unlock()
+	return s.backingStore.RemoveFile(name)
+}
+
+// String describes the backing store for the config.
+func (s *Store) String() string {
+	s.configLock.RLock()
+	defer s.configLock.RUnlock()
```

Comment:
Do we need locking here? The memory store just returns a string, and the file store didn't have a lock originally.

Reply:
I remember adding it for a reason... but now that I look at it, I don't see it. Maybe I removed the code that made it necessary.